---
title: "RNA-seq analysis in R"
author: "Stephane Ballereau, Dominique-Laurent Couturier, Mark Dunning, Abbi Edwards, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
  html_notebook:
    toc: yes
minutes: 300
layout: page
subtitle: Gene Set Testing for RNA-seq
bibliography: ref.bib
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE, cache=FALSE}
library(tidyverse)
load("../Course_Materials/Robjects/Annotated_Results_LvV.RData")
```

```{r setupP, eval=FALSE, cache=FALSE}
library(tidyverse)
# load the data 
load("Robjects/Annotated_Results_LvV.RData")
```

# GO enrichment analysis

## `goseq`

GOseq is a method to conduct Gene Ontology (GO) analysis suitable for RNA-seq 
data as it accounts for the gene length bias in detection of over-representation [[@Young2010]](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14).

From the [GOseq 
vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf):

- GOseq first needs to quantify the length bias present in the dataset under
consideration.
- This is done by calculating a Probability Weighting Function or PWF which can 
be thought of as a function which gives the probability that a gene will be
differentially expressed (DE), based on its length alone.
- The PWF is calculated by fitting a monotonic spline to the binary data series 
of differential expression (1=DE, 0=Not DE) as a function of gene length.
- The PWF is used to weight the chance of selecting each gene when forming a 
null distribution for GO category membership.
- The fact that the PWF is calculated directly from the dataset under 
consideration makes this approach robust, only correcting for the length bias 
present in the data.

```{r goSeqPackage, eval=FALSE, message=FALSE, warning=FALSE}
library(goseq)
supportedOrganisms() %>% filter(str_detect(Genome, "mm"))
```

```{r goSeqPackageTibble, echo=FALSE, message=FALSE, warning=FALSE}
library(goseq)
supportedOrganisms() %>% filter(str_detect(Genome, "mm"))  %>%  
    as_tibble()
```

## Create a list of differentially expressed genes

The input for `goseq` is a vector that indicates, for each gene, whether or not
it is significantly differentially expressed. This should be a named vector, 
where the names are the gene ids and the values are *`1`* if the gene is 
significant and *`0`* if it is not.

In this case we can use the Ensembl gene IDs.

```{r getDEGs}
sigData <- as.integer(!is.na(shrinkLvV$FDR) & shrinkLvV$FDR < 0.01)
names(sigData) <- shrinkLvV$GeneID
```

## Fit the Probability Weighting Function (PWF)

```{r pwFunction}
pwf <- nullp(sigData, "mm10", "ensGene", bias.data = shrinkLvV$medianTxLength)
```

## Conduct GO enrichment analysis

```{r runGoseq, message=FALSE}
goResults <- goseq(pwf, "mm10","ensGene", test.cats=c("GO:BP"))
```

## Plot the top 10

```{r plotGO}
goResults %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count")
```

## Get the GO information for the GO accessions

```{r getGOinfo}
library(GO.db)
GOTERM[[goResults$category[1]]]
```

> ## Challenge 2 {.challenge}
>
> 1. Create a vector showing genes that are statistically significant at 
> FDR < 0.01 and that are up-regulated by at least 4x (logFC>2)
> in lactating mice
> 2. Run a `goseq` analysis on this gene list
> 3. Plot the results
> 4. How is this result different to the previous GO analysis?

```{r solution2, eval=F}

```
